### DISCUSSION TITLE
How to add custom JWKS for id_token ?

### DISCUSSION THREAD

**Discussion by pha3z** [ory/hydra#2599] [Q&A] [answered]
I cannot find anything in the documentation explaining whether or not its possible (and how) to add custom JWKS for id tokens.

I see a lot of doc material on access tokens, but that's not what I'm interested in. Can anyone provide some direction on this?

Additional background:
I have a 3rd party customer who wants to persist the public key used for id_token in their own edge server. They do not want to retrieve the key from my hydra server regularly. Therefore, I need to make sure I can provide them the public key as a one-time configuration deliverable.

I'm not comfortable with the idea of letting Hydra generate the JWKS for me, because I'm paranoid about data loss in the event of a database issue.  In fact, for our infrastructure, I don't even need to persist anything at all in Hydra between restarts, so I've been using in-memory database.  I would like to have this same capability with JWKS for id_token.

**Comment by vinckr** [comment]
You can store and manage your system secret and the generated JWKs somewhere else. 
Maybe this [Terraform Hydra Provider](https://github.com/svrakitin/terraform-provider-hydra) by @svrakitin could help?
I hope this is relevant to your question, let me know if I missed the point 😅 .

**Comment by svrakitin** [comment]
You can overwrite default hydra.openid.id-token key set. https://www.ory.sh/hydra/docs/reference/api/#update-a-json-web-key-set

**Reply by pha3z** [reply]
I find that documentation somewhat confusing.

Is ORY Hydra allowed to store MORE THAN ONE KEY SET?

The documentation for that API call says "Update * * * a * * * JSON Web Key Set" as if there are _multiple key sets_ that you can update. How do I indicate that I want to change the id_token keyset specifically?

**Reply by svrakitin** [reply]
There can be multiple keys within a key set. You can update ID Token keyset to include your custom keypair.

**Reply by pha3z** [reply]
I don't see anything in the example request body or in the documentation that explains how to connect custom keypair to the ID Token keyset.

Regarding PUT /keys/{set}
Example request body shows this:

> {
  "keys": [
    {
      "alg": "RS256",
      "crv": "P-256",
      "d": "T_N8I-...KAE",
      "dp": "G4sP...iBj0",
      "dq": "s9lA...4Q_cgk",
      "e": "AQAB",
      "k": "GawgguFyGrWKav7AX4VKUg",
      "kid": "1603dfe0af8f4596",
      "kty": "RSA",
      "n": "vTqr...0T2IF0",
      "p": "6NbkX...BiJFXQ",
      "q": "0A1Fmp...7vKXAQ",
      "qi": "GyM_p6J...TKhAVRU",
      "use": "sig",
      "x": "f83OJ...vRVEU",
      "x5c": ["string"],
      "y": "x_FEzRu...KIWI5a0"
    }
  ]
}

There's nothing in there that looks like it connects to the id_token.  Maybe I'm dumb, but there's got to be a more complete explanation.

**Reply by pha3z** [reply]
I think I finally found the answer.

[https://www.ory.sh/hydra/docs/guides/secrets-key-rotation](https://www.ory.sh/hydra/docs/guides/secrets-key-rotation)

Quoting from the page:

> ORY Hydra takes the latest key from the key store to sign JSON Web Tokens.

So that leads one to believe that if you add a new key to the key store, the newest key will get used to encrypt all id tokens.  Is that correct?

**Reply by vinckr** [reply]
Yes 🙂.
Maybe @aeneasr can confirm to be really sure. 
I will see to improve the documentation there to make it more clear!

**Reply by pha3z** [reply]
https://www.ory.sh/hydra/docs/reference/api/#update-a-json-web-key-set

It looks like the JWKS is expected to be passed in the URL.  Is it possible to pass in the body instead?

I tried this:

> curl -k -X PUT https://myserver:4445/keys -H "Content-Type: application/json" -d @my_hydra_jwk.json

And I get 404.  I'm guessing because Hydra expects the URL to be https://myserver:4445/keys/{set}.... :(

**Reply by pha3z** [reply]
Erm.  Ok I completely misinterpreted.  The JWKS is supposed to be in the body. 

So what is supposed to go in the {set} part of the URI path?

**Reply by svrakitin** [reply]
hydra.openid.id-token is a name of id token keyset

**Comment by pha3z** [comment, marked as answer]
> curl -k -X GET https://myserver:4445/keys -H "Content-Type: application/json"

Outputs a JWK in this format:

> {"keys":[
{"use":"sig",
"kty":"RSA",
"kid":"public:244a036b-d732-4225-8615-89140574b4b5",
"alg":"RS256",
"n":"...",
"e":"AQAB"},

> {"use":"sig",
"kty":"RSA",
"kid":"private:244a036b-d732-4225-8615-89140574b4b5",
"alg":"RS256",
"n":"...",
"e":"AQAB",
"d":"...",
"p":"...",
"q":"..."
}]}

I generated a public/private key set using this site:
https://mkjwk.org/
But it uses a different format.

I added the generated JWK to ory hydra, but when I visit ".well-known/jwks.json", I receive this error:
> {"error":"error","error_description":"The error is unrecognizable"}

Do I have to use the same format returned from GET?  Or is the problem caused by something else?

**Reply by pha3z** [reply]
I gave up generating JWKS manually. What I did instead:

1) Start Hydra.  Use GET /keys/hydra.openid.id-token to retrieve hydra's generated key.
2) Save the keys to JSON file.
3) Restart Hydra and use PUT /keys/hydra.open.id-token to put the JSON key that I previously saved.

It works.
